add docker build and supporting files

opadron opened the pull request:
Adds a Dockerfile that builds a complete image for running CDash.

Also includes an entrypoint script for the image, with some logic to maintain a statically-defined set of initial users, and a docker-compose.yml file demonstrating how to run the CDash image alongside another image running mysql.

If merged, it would then be pretty easy to point a docker hub account at this repo and automatically deliver new builds with every subsequent merge.

danlamanna commented:
Should just running `docker-compose up` work?

I run into this issue:
```
+ apt-get update
Get:1 http://security.debian.org/debian-security stretch/updates InRelease [94.3 kB]
Ign:2 http://cdn-fastly.deb.debian.org/debian stretch InRelease
Get:3 http://cdn-fastly.deb.debian.org/debian stretch-updates InRelease [91.0 kB]
Get:4 http://security.debian.org/debian-security stretch/updates/main amd64 Packages [374 kB]
Get:5 http://cdn-fastly.deb.debian.org/debian stretch Release [118 kB]
Get:6 http://cdn-fastly.deb.debian.org/debian stretch Release.gpg [2434 B]
Get:7 http://cdn-fastly.deb.debian.org/debian stretch-updates/main amd64 Packages [11.0 kB]
Get:8 http://cdn-fastly.deb.debian.org/debian stretch/main amd64 Packages [7099 kB]
Fetched 7789 kB in 2s (3295 kB/s)
Reading package lists...

## Installing packages required for setup: apt-transport-https lsb-release...

+ apt-get install -y apt-transport-https lsb-release > /dev/null 2>&1

## Confirming "stretch" is supported...

+ curl -sLf -o /dev/null 'https://deb.nodesource.com/node_6.x/dists/stretch/Release'

## Adding the NodeSource signing key to your keyring...

+ curl -s https://deb.nodesource.com/gpgkey/nodesource.gpg.key | apt-key add -
E: gnupg, gnupg2 and gnupg1 do not seem to be installed, but one of them is required for this operation
(23) Failed writing body
Error executing command, exiting
```

opadron commented:
@danlamanna I'm pretty sure someone else has run into that before, but I have no idea what's going on.  The build has always worked perfectly fine for me.

My only guess is maybe you have an old php image?  Did you pull a fresh one prior to building?

Other than that, I've been stumped.

opadron commented:
Ha!  I've finally reproduced this on another host.  I guess I must have been the one working off a stale image somewhere?  Anyway, I think adding gnupg should fix this.

opadron commented:
Hopefully, I've finally nailed this quirk down.  @danlamanna give it another try.

danlamanna commented:
This builds and runs for me now!

Is this meant for development or production? I ask because when I modify files (client side or server side), nothing changes.

opadron commented:
@danlamanna this image is intended for production.  My wish is to see `Kitware/cdash` be the official CDash image on docker hub, but more concretely, I'm working on a project that needs to pull this image from a registry, the most convenient option for which was to have this on docker hub.

bryonbean commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+ && docker-php-ext-configure gd --with-freetype-dir=/usr/include/              \
+                                --with-jpeg-dir=/usr/include/                  \
+ && docker-php-ext-install -j$(nproc) bcmath bz2 gd pdo_mysql pdo_pgsql xsl    \
+ && pecl install xdebug-2.5.5                                                  \
```
@opadron why xdebug if this is for production environments?

opadron replied:
Thanks for catching this.  I'll remove it.

bryonbean commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+FROM php:7.0-apache
+
+MAINTAINER Omar Padron "omar.padron@kitware.com"
```
@opadron Something more generalized like, Kitware "software@kitware.com" is probably more appropriate here but looping in @bradking to enquire about the correct email to use. Also https://docs.docker.com/engine/reference/builder/#maintainer-deprecated.

opadron replied:
I've updated the maintainer section.  I'll adjust further if needed.

bryonbean commented on Dockerfile:
```diff
@@ -0,0 +1,50 @@
+ && docker-php-ext-configure gd --with-freetype-dir=/usr/include/              \
+                                --with-jpeg-dir=/usr/include/                  \
+ && docker-php-ext-install -j$(nproc) bcmath bz2 gd pdo_mysql pdo_pgsql xsl    \
+ && docker-php-ext-enable xdebug                                               \
```
:straggler:

bryonbean replied:
@opadron xdebug is set as a require-dev in the composer.json file, and I didn't realize this so I apologize for overlooking it, but it means that you would also need to use the `--no-dev` flag with composer install, e.g.:
```bash
composer install --no-interaction --no-progress --prefer-dist --no-dev
```

bryonbean replied:
@opadron I'm getting the following error from `docker-compose up --build`

```bash
cdash_1  | /docker-entrypoint.sh: line 2: $'\r': command not found
': not a valid identifieroint.sh: line 3: declare: `__exit_callbacks
cdash_1  | /docker-entrypoint.sh: line 4: syntax error near unexpected token `$'{\r''
'dash_1  | /docker-entrypoint.sh: line 4: `onexit() {
mysql_1  | 180719 22:09:12 [Note] mysqld (mysqld 5.5.60) starting as process 66 ...
mysql_1  | [Entrypoint] Waiting for server...
cdash_cdash_1 exited with code 2
```

opadron commented:
PTAL,.

I've also removed the clone command, and switched to COPY-ing the files from the CWD.

I also try to remove unnecessary files to trim the image as much as possible.  Let me know if there are other things I can remove, if there's a better way for me to do this, or if you think I should go back to cloning (or something else, entirely).

bryonbean commented on Dockerfile:
```diff
@@ -0,0 +1,65 @@
+FROM php:7.0-apache
+
+LABEL maintainer="Kitware, Inc. <software.kitware.com>"
```
@opadron missing @ symbol here.

zackgalbreath replied:
Can we make this `cdash@public.kitware.com`?  That's our public mailing list for CDash, so it's a bit more specific.

opadron commented:
@bryonbean @zachmullen PTAL. I've added a container variable `CDASH_ROOT_ADMIN_EMAIL` (default: `root@docker.container`) as well as `docker.md`, which contains documentation about the available container variables.  Let me know if there's a better place for the documentation.

bryonbean commented:
@opadron Looks good, I think, for the time being at least, the documentation is fine where it's at. +1
